src/Processors/Formats/IOutputFormat.cpp

```diff
@@ -108,7 +105,10 @@ void IOutputFormat::work()
 
 void IOutputFormat::flush()
 {
-    out.next();
+    if (flush_callback)
+        flush_callback(out, result_rows);
```

Comment:
Why this callback needs to pass `out`?
Seems that it is better to pass only `result_rows` and call `out.next()` unconditionally after calling the callback.

Reply:
> and call out.next() unconditionally after calling the callback

But need to clear progress right _before_ flush() and to restore progress right _after_ flush(), this is why I decided to pass `out` inside callback.

Reply:
Ok, I see, thanks!

BTW right now the output is still not clear during query cancellation in clickhouse-local:

```
SELECT *
FROM numbers(1000000000000.)
FORMAT `Null`

Query id: 562caa9f-edd0-42e4-882d-7918edaca718

Cancelling query.
↙ Progress: 1.28 billion rows, 10.20 GB (925.84 million rows/s., 7.41 GB/s.) ▏                                                                                                                                                                                                0%Received exception:
Code: 394. DB::Exception: Query was cancelled. (QUERY_WAS_CANCELLED)
```

But this is how it works in client:

```
SELECT *
FROM numbers(1000000000000.)
FORMAT `Null`

Query id: ba1cc3c0-d27a-4258-b8d3-c9c9d7ee92e3

Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 0.603 sec. Processed 695.27 million rows, 5.56 GB (1.15 billion rows/s., 9.23 GB/s.)
```

So I guess query cancellation should not print an exception in interactive mode in clickhouse-local too

Reply:
BTW progress now also overrides line below (I'm using tmux, and it just overrides the delimiter between panes, not nice)

Reply:
hm, I'm using tmux too but I don't see how it overrides the delimiter between panes, it looks solid line all the time:
![Screenshot from 2021-08-03 12-37-48](https://user-images.githubusercontent.com/54203879/127994166-1056442a-fb05-4542-aaba-6902d87a42fa.png)

Reply:
I've reproduced it only when I have `[layout deed,240x76,0,0[240x38,0,0,75,240x37,0,39,76]] @49` layout (i.e. two panes, splitted horizontally, clickhouse-local in the upper), and in this case it overrides the line below.

I guess it can be tmux issue

Reply:
> (i.e. two panes, splitted horizontally, clickhouse-local in the upper)

Yep, on the photo it is exactly this case - two panes and clickhouse-local in the upper, and it does not get overriden in my tmux.

Reply:
Indeed, it is a tmux issue, which I cannot reproduce with next-3.3 only in 3.1c